docs/reusing-typecheck-results.md

```diff
@@ -0,0 +1,317 @@
+
+## Open questions
+
+### How to cache data?
+
+We can store the data in the intermediate build directory (e.g. `obj` in case of `dotnet build`). The usual way for caching would be to serialize + deserialize information.
+
+However, serialization of the compiler data is non-trivial:
+- we don't have a **full** out-of-the-box solution in the repo (some parts are available, see below)
+- well-known serialization libraries (like `System.Text.Json`) won't work right away, because data is internal and libs can't handle some F# type easily (e.g. DUs)
+- we can't use F#-specific libraries (like [this](https://github.com/fsprojects/FSharp.Json) or [this](https://github.com/Tarmil/FSharp.SystemTextJson)) because they are not "official" enough
+
+It should be possible to hack things around, but the more compiler data we want to serialize, the harder it will be. Hence I see 2 ways to go:
+
+**1) Start with leveraging some existing elements and then see.** 
+
+We have a mechanism for writing signature files. We could force generating signatures for fresh compilation, save them and then "feed" to the recompilation process. This will speed up typechecking, saving a lot of time on type inference. We also seem to have a mechanism for encoding optimization data. We can apply the same idea here - force saving it for the first run and then utilize for reruns. This won't help us with everything but **might be** good enough. Also, the things we'll be left with **might be** easier to hand-serialize.
+
+**2) Create the universal serialization + deserialization mechanism right away and use it for everything**
+
+This can take inspiration from existing mechanisms and random libraries on the internet. Here we'll be able to unlock the full potential of data caching, yet it will also mean a bigger investment in the beginning. Also, the caching procedure would be more uniform here - we can probably store whatever we can per each file, in the same format. 
+
+### Cache invalidation
+
+It's clear that we should be as defensive as possible here, cancelling further information reuse when suspecting that it might not match the fresh recompilation info.
+
+**Checking source files**
+
+To understand the source file is the same we can:
+1) Check the timestamp
+2) Hash the content and compare the hashes
+
+The former might lead to problems with things like time zone changes or manual file metadata modification. The latter might lead to problems in case of hash collisions. All of those scenarios are probably very unlikely, we might just look at what MSBuild does.
+
+**Checking auxiliary information**
+
+Besides tracking the source files, we should track:
+- compiler version
+- compiler flags 
+- ?
+
+**Checking dependencies**
+
+This is related to the minimization of retypecheck in slightly modified projects. If only one file is changed, how do we know that nothing depends on that?
```

Comment:
We don't, that's why we also should save graph TC information. But this might also be not enough, since graph can change even if we save one file. That's why we should for now assume that all files below changed.

Reply:
With graph from last run + graph from new run, you can determine the potential scope of invalidated files (a union of the edges).

Not doing this will significantly decrease usefulness of this features for projects like tests (where the graph has practically no edges).

Reply:
Additionally we should also see how different graphs for projects with signature data and data without it, and see what's the difference will look like for both.

Reply:
Good point about actually caching the TC graph, added.